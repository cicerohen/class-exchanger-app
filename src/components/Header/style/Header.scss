@import "./Header.variables";

.root {
  padding-left: $defaultPadding;
  padding-right: $defaultPadding;
  background-color: $headerBgColor;
}

.container {
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  height: 80px;
  align-items: center;
}

.contentLeftContainer {
  display: flex;
}

.contentRightContainer {
  display: flex;
}

.menu__root {
  composes: root from 'components/Menu/style/Menu.scss';
  display: flex;
}

.menuItem__container {
  composes: container from 'components/MenuItem/style/MenuItem.scss';
  color: $headerColor;
}