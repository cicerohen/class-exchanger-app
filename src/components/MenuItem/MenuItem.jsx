import React from 'react';
import PropTypes from 'prop-types';
import Style from './style/MenuItem.scss';

const MenuItem = ({
  children, href, title, onClick, style,
}) => (
  <li className={style.root}>
    <a className={style.container} href={href} title={title} onClick={onClick}>
      { children }
    </a>
  </li>
);

MenuItem.defaultProps = {
  href: '#',
  onClick: undefined,
  title: undefined,
  style: Style,
};

MenuItem.propTypes = {
  href: PropTypes.string,
  title: PropTypes.string,
  onClick: PropTypes.func,
  style: PropTypes.objectOf(PropTypes.any),
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
};

export default MenuItem;
