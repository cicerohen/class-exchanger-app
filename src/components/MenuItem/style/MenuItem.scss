@import './MenuItem.variables';

.root {
  list-style: none;
}

.container {
  text-decoration: none;
  padding-left: 15px;
  padding-top: 10px;
  padding-right: 15px;
  padding-bottom: 10px;
  display: block;
}
