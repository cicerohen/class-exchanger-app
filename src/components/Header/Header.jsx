import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FaBars } from 'react-icons/lib/fa';
import Menu from '../Menu';
import MenuItem from '../MenuItem';
import menuStyle from '../Menu/style/Menu.scss';
import menuItemStyle from '../MenuItem/style/MenuItem.scss';

import Style from './style/Header.scss';

class Header extends Component {
  goToMe = (e) => {
    const { pushFn } = this.props;
    e.preventDefault();
    pushFn('me');
  }

  goToHome = (e) => {
    const { pushFn } = this.props;
    e.preventDefault();
    pushFn('/');
  }

  goToSignIn = (e) => {
    const { pushFn } = this.props;
    e.preventDefault();
    pushFn('signin');
  }

  goToSignUp = (e) => {
    const { pushFn } = this.props;
    e.preventDefault();
    pushFn('signup');
  }

  signOut = (e) => {
    const { signOutFn } = this.props;
    e.preventDefault();
    signOutFn();
  }

  openSidebar = (e) => {
    const { openSidebarFn } = this.props;
    e.preventDefault();
    openSidebarFn();
  }

  render() {
    const { isAuthenticated } = this.props;
    return (
      <header className={Style.root}>
        <div className={Style.container}>
          <div className={Style.contentLeftContainer}>
            <Menu style={{ ...menuStyle, root: Style.menu__root }}>
              <MenuItem
                onClick={this.openSidebar}
                style={{ ...menuItemStyle, container: Style.menuItem__container }}
              >
                <FaBars />
              </MenuItem>
              <MenuItem
                href="/"
                onClick={this.goToHome}
                style={{ ...menuItemStyle, container: Style.menuItem__container }}
              >
                Home
              </MenuItem>
            </Menu>
          </div>
          <div className={Style.contentRightContainer}>
            {
              isAuthenticated &&
              <MenuItem
                href="me"
                onClick={this.goToMe}
                style={{ ...menuItemStyle, container: Style.menuItem__container }}
              >
                Me
              </MenuItem>
            }
            {
              isAuthenticated &&
              <MenuItem
                onClick={this.signOut}
                style={{ ...menuItemStyle, container: Style.menuItem__container }}
              >
                Sign Out
              </MenuItem>
            }
            {
              !isAuthenticated &&
              <MenuItem
                href="signin"
                onClick={this.goToSignIn}
                style={{ ...menuItemStyle, container: Style.menuItem__container }}
              >
                Sign In
              </MenuItem>
            }
            {
              !isAuthenticated &&
              <MenuItem
                href="signup"
                onClick={this.goToSignUp}
                style={{ ...menuItemStyle, container: Style.menuItem__container }}
              >
                Sign Up
              </MenuItem>
            }
          </div>
        </div>
      </header>
    );
  }
}

Header.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  pushFn: PropTypes.func.isRequired,
  openSidebarFn: PropTypes.func.isRequired,
  signOutFn: PropTypes.func.isRequired,
};

export default Header;
