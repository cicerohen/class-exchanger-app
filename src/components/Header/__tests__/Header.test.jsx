import React from 'react';
import { shallow } from 'enzyme';
import { FaBars } from 'react-icons/lib/fa';
import Header from '../Header';
import MenuItem from '../../MenuItem';

const eventMockObject = {
  preventDefault: jest.fn(),
};

const minimalProps = {
  isAuthenticated: false,
  pushFn: jest.fn(),
  openSidebarFn: jest.fn(),
  signOutFn: jest.fn(),
};

describe('<Header /> Tests', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render without exploding', () => {
    const wrapper = shallow(<Header {...minimalProps} />);
    expect(wrapper).toHaveLength(1);
  });

  it('should render correctly with minimal props', () => {
    const wrapper = shallow(<Header {...minimalProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctlys when is authenticated', () => {
    const props = {
      ...minimalProps,
      isAuthenticated: true,
    };
    const wrapper = shallow(<Header {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call pushFn when click on Me menu item', (done) => {
    const props = {
      ...minimalProps,
      isAuthenticated: true,
    };
    const wrapper = shallow(<Header {...props} />);
    const menuItem = wrapper.find(MenuItem).filterWhere(item => item.children().contains('Me'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
    expect(minimalProps.pushFn.mock.calls.length).toBe(1);
    done();
  });

  it('should call pushFn when click on Home menu item', (done) => {
    const wrapper = shallow(<Header {...minimalProps} />);
    const menuItem = wrapper.find(MenuItem).filterWhere(item => item.children().contains('Home'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
    expect(minimalProps.pushFn.mock.calls.length).toBe(1);
    done();
  });

  it('should call pushFn when click on Sign In menu item', (done) => {
    const wrapper = shallow(<Header {...minimalProps} />);
    const menuItem = wrapper.find(MenuItem).filterWhere(item => item.children().contains('Sign In'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
    expect(minimalProps.pushFn.mock.calls.length).toBe(1);
    done();
  });

  it('should call pushFn when click on Sign Up menu item', (done) => {
    const wrapper = shallow(<Header {...minimalProps} />);
    const menuItem = wrapper.find(MenuItem).filterWhere(item => item.children().contains('Sign Up'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
    expect(minimalProps.pushFn.mock.calls.length).toBe(1);
    done();
  });

  it('should call signOutFn when click on Sign Out menu item', (done) => {
    const props = {
      ...minimalProps,
      isAuthenticated: true,
    };
    const wrapper = shallow(<Header {...props} />);
    const menuItem = wrapper.find(MenuItem).filterWhere(item => item.children().contains('Sign Out'));
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
    expect(minimalProps.signOutFn.mock.calls.length).toBe(1);
    done();
  });

  it('should call openSidebarFn when click on hamburguer menu item', (done) => {
    const wrapper = shallow(<Header {...minimalProps} />);
    const menuItem = wrapper.find(MenuItem).filterWhere(item => item.children().type() === FaBars);
    const menuItemAnchor = menuItem.dive().shallow().find('a').first();
    menuItemAnchor.simulate('click', eventMockObject);
    expect(minimalProps.openSidebarFn.mock.calls.length).toBe(1);
    done();
  });
});
