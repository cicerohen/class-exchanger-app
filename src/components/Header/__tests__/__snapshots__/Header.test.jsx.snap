// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> Tests should render correctly with minimal props 1`] = `
<header
  className="root"
>
  <div
    className="container"
  >
    <div
      className="contentLeftContainer"
    >
      <Menu
        style={
          Object {
            "root": "menu__root",
          }
        }
      >
        <MenuItem
          href="#"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          <FaBars />
        </MenuItem>
        <MenuItem
          href="/"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          Home
        </MenuItem>
      </Menu>
    </div>
    <div
      className="contentRightContainer"
    >
      <MenuItem
        href="signin"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Sign In
      </MenuItem>
      <MenuItem
        href="signup"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Sign Up
      </MenuItem>
    </div>
  </div>
</header>
`;

exports[`<Header /> Tests should render correctlys when is authenticated 1`] = `
<header
  className="root"
>
  <div
    className="container"
  >
    <div
      className="contentLeftContainer"
    >
      <Menu
        style={
          Object {
            "root": "menu__root",
          }
        }
      >
        <MenuItem
          href="#"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          <FaBars />
        </MenuItem>
        <MenuItem
          href="/"
          onClick={[Function]}
          style={
            Object {
              "container": "menuItem__container",
            }
          }
        >
          Home
        </MenuItem>
      </Menu>
    </div>
    <div
      className="contentRightContainer"
    >
      <MenuItem
        href="me"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Me
      </MenuItem>
      <MenuItem
        href="#"
        onClick={[Function]}
        style={
          Object {
            "container": "menuItem__container",
          }
        }
      >
        Sign Out
      </MenuItem>
    </div>
  </div>
</header>
`;
